experimental client: Add MetadataBundle

jku opened the pull request:
MetadataBundle keeps track of current valid set of metadata for the
client, and handles almost every step of the "Detailed client workflow"
in the TUF specification (the remaining steps are I/O related).

It also
verifies any new metadata (local or downloaded from remote repository).

**EDIT** this bit is no longer true in the newest version: _The bundle takes care of persisting valid metadata on disk, loading
local metadata from disk and deleting invalid local metadata._

Commit:
experimental client: Add MetadataBundle

MetadataBundle keeps track of current valid set of metadata for the
client, and handles almost every step of the "Detailed client workflow"
in the TUF specification (the remaining steps are download related).

The bundle takes care of persisting valid metadata on disk, loading
local metadata from disk and deleting invalid local metadata. It also
verifies any new metadata (downloaded from remote repository) it is
given.

This is very much a work-in-progress.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Update outdated docstring

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
From discussion with @sechkova: We do need to expose local metadata loading (instead of doing it automatically) for two reasons:
* delegate handling is going to need it
* spec "requires" downloading top-level metadata but in fact we can avoid it if we know the current local metadata is valid (valid for snapshot and targets I suppose)

For delegate handle we also need to have an optional delegator role.

So instead of just `update_metadata(metadata: str)` we'll likely have something like
```
def load_local_metadata(role: str, delegator_role: Optional[str]) -> bool
    """Tries to load 'role' metadata from disk
    Returns True if valid, up-to-date metadata was found and loaded
    """

def update_metadata(metadata: str, delegator_role: Optional[str])
    """Loads and persists given metadata on disk
    Raises if metadata is not valid or could not be verified
    """
```

Commit:
MetadataBundle: Modify state handling

A single state variable cannot really handle the case where we may want
load local metadata and may want to updater from remote -- but are not
required to do either.

Instead, make sure two rules apply:
* Metadata can only be loaded/updated if all metadata it depends on is
  loaded
* Metadata becomes immutable when anything that depends on it is loaded

So, loading a new timestamp is possible if root is loaded and snapshot
is not loaded.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: implement delegates support

Also make sure we always try loading local data for the metadata
that gets security benefits from it.

Also update the threshold verification (placeholder) to support
targets as well as root.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: use Metadata.from_bytes()

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Require load_local_metadata()

Change the rules a bit: require calling load_local_metadata() before
update_metadata() is legal. This removes the need to run
load_local_metadata() "just to be sure" in update_metadata()

Improve comments

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
Initial draft was very neat and clean but after some impacts with reality (thanks Teodora for early review), it currently looks like this:
 * ~50 lines of code that should in the end be in metadata
 * 160 lines of code for internal state tracking (so e.g. root can't be updated when targets is already loaded), and quality of life improvements like properties
 * ~160 lines of code for the actual spec checks (these _should_ be complete now)
 
I still think this is a good direction. The code is not ready by any means but I'm making this pr now so that Teodora can start using it early if we decide this is a path experimental-client should take. I suggest reviewing the final form and not individual commits -- there were false starts in there.

The second bullet point in the list above is a potential issue: is the internal state tracking understandable and clear (or can it be improved to be that).

The other potential issue is delegated metadata. Delegates can be added and they are validated on insert but they are not currently under the same guarantees that top-level metadata is: if there's a top-level "targets" in the bundle it's guaranteed to be valid, but nothing currently stops caller from updating  "role1" so that it no longer delegates to "role2" that is also in the bundle... This can definitely be fixed (I think just reloading all metadata delegated by the just updated metadata should do it and might even simplify the internal state tracking : I just haven't had time to try yet).

Third things to mention might be exceptions: I believe one is thrown where needed now... but the actual exceptions are not chosen at all: My plan is that all issues that arise from metadata content will be under one base Exception. Client will likely just handle the base exception. Other exceptions will be internal errors and user errors.

jku commented:
On the API and naming:

```
MetadataBundle
  # Implements Mapping so individual metadata can be access with []-indexing

  # methods
  def load_local_metadata(...)
  def update_metadata(...)
  def root_update_finished(...)

  # helper properties
  root
  snapshot
  timestamp
  targets
```
MetadataBundle itself is not a great name but works -- I know Teodora is fond of "trusted" or some variation of it in the name to make that part clear.

the loader function names could be `load_local()`, `update_from_remote()` to make the distinction clear but names shorter.
 `root_update_finished(...)` is not actually strictly speaking required -- the root verification could safely be made on first timestamp load -- but I've kept it there since it matches the specification better (it's the marker for root update loop having finished).

MVrachev commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,349 @@
+
+
+def from_string(data: str) -> Metadata:
+    return JSONDeserializer().deserialize(data)
```
You can start using `Metadata.from_bytes()` after it was added by you, right?

MVrachev commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,349 @@
+
+
+class MetadataBundle(abc.Mapping):
+    def __init__(self, path: str):
```
Nitpick: Maybe it's a good idea to use a little more descriptive name here.

It's a little unclear: is this path to specific metadata and all others are in the same folder or it's a path to the parent folder of all expected metadata files?
Probably something like `repository_path` or if you want to keep it short: `repo_path`.

MVrachev commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,349 @@
+        self._state = "root"
+        self.reference_time = None
+
+        if not os.path.exists(path):
```
Nitpick: I would expect function argument checks to be before attribute assignment.

jku replied:
On second thought it's not actually _argument_ validation: we're checking whether the path exists on filesystem and might even create it in this code path... so I think the order is fine

MVrachev commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,349 @@
+        self.reference_time = None
+
+        if not os.path.exists(path):
+            # TODO try to create dir instead?
```
Yes, I will expect that to happen if there is no repository in the path I have passed.

jku replied:
Since bundle does not support any "root bootstrapping" at this point creating directories would be pointless so I removed this whole check

MVrachev commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,450 @@
+        keys = delegator.signed.delegations["keys"]
+        # role names are unique: first match is enough
+        roles = delegator.signed.delegations["roles"]
+        role = next((role for role in roles if role["name"] == role_name), None)
```
Not sure why do you take the `next` one?
Don't you want exactly the role which equals the `role_name`?

jku replied:
I'm looking for the role in roles that has the name `role_name`.

the list comprehension works as a filter that is also an iterator, and next() gives me the first value from that iterator (or None if no match).

API could maybe offer `Targets.get_delegation(role_name)` -- but let's not move too hastily: this calling function should IMO live in the API itself (see #1306).

MVrachev commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,450 @@
+   by same immutability guarantees: the top-level metadata is handled
+   by hard-coded rules (can't update root if snapshot is loaded)
+   but delegations would require storing the delegation tree ...
+ * exceptions are all over the place and not thought out at all
```
I would add  `Update docstrings with what exceptions are raised`, so that you won't forget.
Right now, you are documenting `Raises if X` in the functions docstrings.

MVrachev commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,450 @@
+                new_timestamp.signed.meta["snapshot.json"]["version"]
+                < self.timestamp.signed.meta["snapshot.json"]["version"]
+            ):
+                # TODO not sure about the
```
Do you remember what you meant here? 
Probably not sure about the exception?

jku replied:
yes almost certainly :)

MVrachev commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,450 @@
+   by hard-coded rules (can't update root if snapshot is loaded)
+   but delegations would require storing the delegation tree ...
+ * exceptions are all over the place and not thought out at all
+ * usefulness of root_update_finished() can be debated: it could be done
```
Yea, I agree and I don't see where are you using it now?

sechkova replied:
Updater calls it, it is a public method.
 If we want to be strict with the spec and do what it calls "protection from fast forward attacks" , I think this is the place. Although I admit, you already got me here because I forgot to call `root_update_finished()` and got an error :)

jku replied:
it's a public method that client should currently call after final root has been loaded -- according to spec we verify root expiry at this point... Practically we could verify it when the next metadata (timestamp) is about to be loaded but it looks a bit goofy: caller is trying to load timestamp but gets an error saying root is suddenly not valid.

MVrachev reviewed (changes requested):
Overall, the code is easy to read and follow. The few complicated places (like verifying hashes) have issues that those things can be added in the Metadata API.
Also, I agree that helper functions could be useful like `get_snap_version()` in timestamp.

One additional suggestion.
Right now, you are calling `self._raise_on_unsupported_state(role_name)` when loading and updating, and you do checks at the beginning of each of `_load_*` functions.
That way it's easy to forget or duplicate checks.

I would suggest using `self._raise_on_unsupported_state(role_name)` to handle all state checks
and call it at the begging of each of the `_load_*` functions.

jku commented:
> Right now, you are calling `self._raise_on_unsupported_state(role_name)` when loading and updating, and you do checks at the beginning of each of `_load_*` functions.
> That way it's easy to forget or duplicate checks.

Yeah this is true

> I would suggest using `self._raise_on_unsupported_state(role_name)` to handle all state checks
> and call it at the begging of each of the `_load_*` functions.

I think I want to only do it early in the public functions and am 99% sure I can just remove the checks in the actual `_load_*()` methods... just haven't gone through the trouble of verifying that yet.

Commit:
Improve loading rules checks

Make sure we do not load any metadata if delegates of
that metadata are already loaded.

Also remove duplicate checks.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Comment and variable name improvements

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

trishankatdatadog commented:
There's a lot of if-else inside generic-sounding functions. Is that a good idea? Should we perhaps use specific functions for specific purposes? Not only improves readability, but also detecting errors.

jku commented:
> There's a lot of if-else inside generic-sounding functions. Is that a good idea? Should we perhaps use specific functions for specific purposes? 

You're totally correct (I debated myself about writing about this but ... I wrote so many comments already :) ).

Let me drill down into this a bit. There are two classes of if-elses:
* state checks in _raise_on_unsupported_state(): this does not so much make program flow decisions but errors out if a logical error (a programming error) is about to happen. This code can possibly be simplified but I argue it has its place... It could also be removed completely, if we just assume caller never makes mistakes :)
* The other case is the public functions `load_local_metadata()` and `update_metadata()` that operate as dispatchers to the specific internal functions -- here I completely agree with you that specific functions would be better (instead of a dispatcher we would have more public methods like `load_local_root`, `load_local_timestamp` etc that would be self explanatory and harder to make mistakes with). However this ties to the mirrors discussion (more below)

This relates to mirrors in the following way:
* we recognised the download/mirrors code is what makes this tricky -- this is what Teodora tried to solve in some earlier PRs but ultimately failed to create something that improves on the current code _and still supports mirrors_...
* as long as the mirror downloading works as it does, it means there is going to be a "generic download loop" that is going to download files, parse the content, verify the metatadata and then decide if it needs to download another version of the file. Anything else is going to lead to a lot of repetition in the updater code.
* This means that if bundle does not offer generic dispatcher functions, then the calling updater code is going to have to do the exact same thing instead

So I believe the dispatching code (the if-clauses that look at role name to decide what to call) is going to exist _somewhere_ anyway (again, assuming current mirroring support)


So I guess this brings us to: **Re-designing mirror support (essentially limiting it drastically) would make this, and the updater download code, much cleaner**. We've been really trying to not solve too many issues at once but maybe we do need to look at that as well...


(marking draft until the mirrors discussion happens)

Commit:
Exceptions refactoring

Make more exceptions derive from RepositoryError:
The idea is that a client can just handle RepositoryError and will know
that any problems resulting from unexpected metadata will end up there.

Also fix some wildly misleading variable naming in ReplayedMetadataError

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Exception refactor

The goal is that the bundle only raises two kinds of errors:
* user errors (ValueError/KeyError) that can be avoided by
  the caller
* RepositoryErrors that are a result of unacceptable metadata:
  The requested action cannot succeed because of the metadata.
  These typically cannot be avoided by the caller.

File open and serialization errors are handled internally.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Update MetadataBundle with named methods

Use named methods for the top-level metadata (e.g.
load_local_timestamp() instead of load_local_metadata("timestamp")).
This can now happen after the mirrors are no longer in play (so
updater can easily call named methods) and makes the bundle
implementation more straight-forward.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Handle targets like delegated targets

The public load_local_targets() and update_targets() can just call the
versions for delegated targets to remove duplication.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Avoid loading targets twice

If role is loaded already, skip loading local version (this is an
optimization since there's no case where the loaded version and the
on-disk version should differ).

This could be done for top-level metadata as well but the situation
doesn't really come up there: there's no good reasons to try loading
top-level local metadata multiple times, unlike in the targets case
where same delegates may be loaded when looking up different target
files.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: fix import name

Also use only lookup dictionary once if necessary.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Improve hints and docs

Complete the type hints for MetadataBundle.
Slightly improve documentation.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Remove unnecessary directory check at startup

* Loading root.json will fail just as descriptively
* As long as Bundle doesn't implement bootstrapping
  the local repo, there's also no need to create
  missing directories

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
Updating this based on the mirrors/downloads changes:
* functions modifying top-level metadata are now named so `update_root(data)` instead of `update_metadata(data, "root")`
* all of the type-dispatching was removed as result
* minimal internal state checking is now done in `_load_*()` (the functions that actually do the spec checks)

I'm quite happy with this: it's not any shorter than the previous version but it's simple and straight forward. I've tested using the bundle from the client and mostly it feels like a clear improvement: the client code can concentrate on deciding what should happen (_"Now load snapshot from local storage: if that one's not valid anymore, then load snapshot from remote"_) while the bundle verifies that metadata state is valid according to spec at every step.

It's definitely not complete (most importantly missing tests and exceptions are still WIP, and the spec validation needs to be reviewed carefully by someone who's not me) but I think it might be good enough for the experimental-client? Opinions as well as naming suggestions are welcome. On naming: is MetadataBundle ok? load_local_*()? update_*()?

TLDR documentation:
> MetadataBundle is an internal class for the Updater. It's purpose is to keep track of valid metadata, ensure that this metadata is always valid, and to provide easy ways to update the metadata with the caller making decisions on what is updated.
>
> ## Modifying metadata
> `load_local_*()`: Load metadata from local cache
> `update_*()`: Update and persist metadata using data from remote repository
> `root_update_finished()`: Mark root metadata as final (to verify it's not expired)
> 
> ## Accessing loaded metadata
> Use the Mapping index access: `bundle['root']`, `bundle.get('root')`
> For top-level metadata use the helpers: `bundle.root`
> 
> ## Internal
> `_load_*()`: Actual spec checks, these are called by both `load_local_*()` and `update_*()`

(EDIT: rebasing this to top of the branch brings out more linting issues: I'll still be fixing those)

jku commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,482 @@
+     the load functions will refuse to load the files when they are not signed by
+     current root keys. Deleting at the specified point is possible but means additional
+     code with some quirks..
+   * in general local metadata files are not deleted (they just won't succesfully load)
```
This might be worth a discussion. I don't like deleting local metadata files when they're invalid for two reasons:
* it doesn't provide security (if the metadata is invalid it cannot be loaded -- if it could then our program has a bug that should be fixed: deleting the metadata won't really help)
* it prevents all debugging: "oh the app says update failed because of invalid metadata... well, that file doesn't exist anymore :shrug:"

MVrachev replied:
Wondering is there a possible situation where you can fill up the storage by providing let's say thousands of invalid metadata files? 

Other than that your arguments are sensible.

Commit:
MetadataBundle: Fix loads of linting issues

Lots of fixes, mostly obvious ones. The trickier ones and pylint
disables have comments added to explain them.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
Had to rebase on current experimental-client to get the linting up-to-date: of course the newer linting revealed lots of new issues that I've fixed or documented :)

Even more than before I suggest not reading individual commits: there's a lot of less-than-useful history there.

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,505 @@
+# pylint: disable=logging-too-many-args
+
+# TODO: signed._type really does not work issue #1375:
+# pylint: disable=protected-access
```
For the whole module? 😱 Let's rebase after the merge of #1376

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,505 @@
+def verify_with_threshold(
+    delegator: Metadata, role_name: str, unverified: Metadata
+) -> bool:
+    """Verify 'unverified' with keys and treshold defined in delegator"""
```
```suggestion
    """Verify 'unverified' with keys and threshold defined in delegator"""
```

sechkova commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,505 @@
+                f"Expected 'root', got '{new_root.signed._type}'"
+            )
+
+        if self.root is not None:
```
A comment will be useful to show that `if root is None` means that we are loading the local trusted root, otherwise we are updating it.

sechkova commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,505 @@
+        if self.root.signed.is_expired(self.reference_time):
+            raise exceptions.ExpiredMetadataError("New root.json is expired")
+
+        logger.debug("Verified final root.json")
```
Not sure what would be the right place to put this but I think we need a text somewhere saying that the by design the bundle avoids fast-forward attacks (it does not load previous snapshot and timestamp on init) and does not need to perform explicitly step 5.3.11 from the spec.

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,505 @@
+"""TUF client bundle-of-metadata
+
+MetadataBundle keeps track of current valid set of metadata for the client,
+and handles almost every step of the "Detailed client workflow" in the TUF
```
Worth adding a link? https://theupdateframework.github.io/specification/latest/#detailed-client-workflow

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,505 @@
+ * Metadata is immutable if any metadata depending on it has been loaded
+ * Caller must load/update these in order:
+   root -> timestamp -> snapshot -> targets -> (other delegated targets)
+ * Caller should try loading local file before updating metadata from remote
```
Why? What happens if they don't? We should somehow distinguish, at least in our documentation, between initial/trusted root metadata – which [must exist](https://theupdateframework.github.io/specification/latest/#load-trusted-root) – and other local metadata.

jku replied:
Root is handled separately: local root load happens during instance initialization. I should make that clear in this text.

For all others I wanted to leave the decision to caller: it would be possible to keep track of local loads within the bundle and to do it automatically when needed, but it's more code and the caller is going to do it anyway (because in several cases caller wants to know if the local metadata is valid before it decides whether to even try downloading remote metadata)...

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,505 @@
+>>>
+>>> # load snapshot, update from remote if needed
+>>> if not bundle.load_local_snapshot():
+>>>     # TODO get version from timestamp
```
For this and the other TODO entries in the example, are we waiting for the `MetadataInfo` class to merge #1329?

jku replied:
it's just that currently it ends up being about five lines of code more which makes the example a bit heavy... This is roughly what my test branch looks like:

```
        # load local snapshot, update from remote only if local is not valid
        if not self._bundle.load_local_snapshot():
            # Use length if available, use versioned name if consistent_snapshot
            filename = "snapshot.json"
            metainfo = self._bundle.timestamp.signed.meta[filename]
            length = metainfo.get("length") or DEFAULT_SNAPSHOT_MAX_LENGTH
            if self._bundle.root.signed.consistent_snapshot:
                filename = f"{metainfo['version']}.{filename}"

            data = self._download_metadata(filename, length)
            self._bundle.update_snapshot(data)
``` 

Maybe TODO is the wrong word: I should just mention that the example doesn't use consistent_snapshot

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,505 @@
+
+# TODO: Either enaable old-style logging in pylintc (issue #1334)
+# or change this file to use f-strings for logging
+# pylint: disable=logging-too-many-args
```
Per discussion in #1334, let's enable "old-style" logging.

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,505 @@
+        logger.debug("Updating root")
+
+        self._load_intermediate_root(data)
+        self.root.to_file(os.path.join(self._path, "root.json"))
```
Should we be persisting intermediate root metadata with its versioned prefix? Not necessarily from a spec compliance perspective, but more in-line with your argument about not deleting files as it aides debugging.

jku replied:
IMO spec compliance explicitly does not require this, btw: the part that talks about writing consistent snapshots seems to talk about the _remote_ repository to me, and the part that talks about client writes in general explicitly says files must be written as FILENAME.EXT.

Now, we will likely want to write versioned root files at some point anyway (to handle the bootstrap case) but I'm fine with adding that later on and keeping the code as simple as possible for now.

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,505 @@
+        # Store our reference "now", verify root expiry
+        self.reference_time = datetime.utcnow()
+        if self.root.signed.is_expired(self.reference_time):
+            raise exceptions.ExpiredMetadataError("New root.json is expired")
```
Neat trick. The specification suggests the update start time before updating root. Arguably we should assign `self.reference_time` earlier in the `MetadataBundle` logic. What do you think?

jku replied:
Maybe yes. The spec does say that...

This choice made sense to me as it is (very close to) the time that client last checks for root.json -- so theoretically  there's actual value in comparing it to e.g. server publish times if issues are debugged...

But this could be a situation where following the spec to letter is a better choice. That just means I need an extra variable for `root_is_final` :)

sechkova replied:
I wanted to suggest adding an additional flag marking `root_update_finished`. I see that reference time serves exactly like that but something like `if root_update_finished:` is instantly comprehensible.

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,505 @@
+    def load_local_timestamp(self) -> bool:
+        """Load cached timestamp metadata from local storage.
+
+        Returns True if timestamp was succesfully loaded"""
```
This was a surprise. Especially as there were not any signs of checking return type of the `load_local_*` functions in the example snippet near the top of the file. Wouldn't the Pythonic way to do this be to throw an exception if the `load_local_*()` functions do not succeed? Or was this intentional, because only root metadata **must** exist locally?

jku replied:
Examples do do things like `if not bundle.load_local_snapshot()`.

This was intentional design ... but I'm not sure it was correct. The logic is that failing to load non-root local metadata for any reason (it does not exist, it is expired, it is no longer signed by correct keys, ...) is completely normal for the caller: in all cases the caller can only assume things are working as expected and that remote metadata should now be downloaded.

Would it be more pythonic to raise? I guess so, it doesn't look horrible anyway (simplified example but still...):

```python
try:
    bundle.local_local_timestamp()
except: LocalLoadError as e:
    logger.info("Failed to load local timestamp: %s", e)

# Always load remote timestamp, even if local is valid
data = download_metadata(timestamp_filename)
bundle.update_timestamp(data)

try:
    bundle.local_local_snapshot()
except: LocalLoadError as e:
    logger.info("Failed to load local snapshot: %s", e)

    # Load remote snapshot only if local is not valid
    data = download_metadata(snapshot_filename)
    bundle.update_snapshot(data)
```

Maybe I need to try this

jku replied:
Exceptions work fine, I'll use them throughout

joshuagl reviewed (commented):
This is very nice, easy to follow even when attention is depleted. Nice work @jku !

The class level docstring suggests that invalid local metadata is deleted. I think you made a conscious decision _not_ to do that? It looks like the current implementation only ever preserves the most recently verified metadata file for each role? If we did start storing intermediate files on disk, we would need to add a mechanism to prevent the local metadata directory getting too large. But I don't think that's necessary in the current state.

There's a lot of LBYL style code `load_local_*`, `is_expired()` and `if not verify_with_threshold()`, whereas I think we want to lean into the Pythonic EAFP. Can we reasonably have each of those, especially the latter two, raise Exceptions that would be reasonable to surface to the user? Or would we need to catch them and re-raise to provide meaningful errors?

It will be nice to see this without the repeated logic, such as the big block for checking hashes in meta.

Also, this appears to be internal API – did we figure the best way to indicate that? Or is that TBD?

jku commented:
Thanks, good to get "fresh" eyes on this.

> The class level docstring suggests that invalid local metadata is deleted.

Oops.

> There's a lot of LBYL style code load_local_*, is_expired() and if not verify_with_threshold(), whereas I think we want to lean into the Pythonic EAFP. Can we reasonably have each of those, especially the latter two, raise Exceptions that would be reasonable to surface to the user? Or would we need to catch them and re-raise to provide meaningful errors?

I've gone into detail about load_local_* in a code comment (you may be right).  `is_expired()` and `verify_with_threshold()` are Metadata API questions (yes the latter is currently implemented here but it should be in Metadata API in my opinion): I don't object to exceptions but suggest we handle them elsewhere: I recently filed #1377 (which I see you've already seen).

> It will be nice to see this without the repeated logic, such as the big block for checking hashes in meta.

Yes, hash checks should also be part of Metadata API (#1377) in my opinion.

> Also, this appears to be internal API – did we figure the best way to indicate that? Or is that TBD?

My suggestion was to move everything except updater.py and fetcherinterface.py into a tuf/client/_internal/ directory. Would make sense to do before we merge to develop.

sechkova commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,505 @@
+        Returns True if metadata was succesfully loaded"""
+        if self.get(role_name):
+            logger.debug("Local %s already loaded", role_name)
+            return True
```
Why is this lookup done only for targets? I guess the intention is making the delegations tree traversal more efficient?

jku replied:
Yeah it's true we could do it for others too -- it's just that the issue should not come up for top-level metadata but definitely can come up for delegated targets.

sechkova reviewed (commented):
I have one doubt about file system access. 
On one hand, by doing it inside the bundle, as it is now, the bundle itself is in control of persisting only verified metadata files that passed all checks which seems less error prone. Also keeps the `Updater` code simpler.

But on the other hand, if the caller (`Updater`) handles both access to the file system and the network download, MetadataBundle will implement only the actual metadata verification steps from the client workflow, will be I/O agnostic and will provide a single public method per metadata type (basically the now private `_load_*` methods).

Other than that, I tried to follow the specification and I think all steps that do not relate to the network download are implemented (besides consistent_snapshot and some TODOs already in the code). But for sure another look is needed when all client components are combined together.

jku commented:
> But on the other hand, if the caller (`Updater`) handles both access to the file system and the network download, MetadataBundle will implement only the actual metadata verification steps from the client workflow, will be I/O agnostic and will provide a single public method per metadata type (basically the now private `_load_*` methods).

These are valid points. I guess the main idea was to get (almost) all spec steps into the same file so it would be easy to follow. But I can definitely be persuaded on this.

Commit:
Improve documentation

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

sechkova commented:
> > But on the other hand, if the caller (`Updater`) handles both access to the file system and the network download, MetadataBundle will implement only the actual metadata verification steps from the client workflow, will be I/O agnostic and will provide a single public method per metadata type (basically the now private `_load_*` methods).
> 
> These are valid points. I guess the main idea was to get (almost) all spec steps into the same file so it would be easy to follow. But I can definitely be persuaded on this.

We can come back to this idea when the different PRs are pieced together.

Commit:
MetadataBundle: Update to API changes

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
tests: Add minimal test case for Bundle

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Metadata API: Fix DelegatedRole serialization issue

A DelegatedRole with paths=[] fails to serialize correctly (paths is not
included in the output json).

Fix the issue, modify tests to notice a regression.

Fixes #1389

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Metadata API: Don't do equality comparisons on containers

Use either "if X is not None:" or a try-except instead of a "if X:".

I believe Targets.from_dict() was not really broken with previous code
but it looks suspicious and did fail the added test with a strange
exception: I expect the from_dict() methods to mainly fail with
KeyErrors, ValueErrors or AttributeErrors if file format structure
is incorrect.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Change ValueErrors to RuntimeErrors

As the metadata type is no longer an argument, these are not
ValueErrors.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Save original files on disk

Don't use the serialized format as that won't match any hashes in
"meta".

Add basic tests for updating metadata.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Store reference time earlier

Spec says reference time should be the beginning of the process: do
that.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Use type, not _type

Signed now has "type" attribute, use that.

Also remove another pylint disable that is no longer needed (logging is
now old style).

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Add comments about the process

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Handle Deserialization errors

because we are deserializing, not serializing.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Raise instead of returning bool

The bundle should now raise
* derivatives of RepositoryError on failures that are likely a result
  of server error or a malicious server
* RuntimeErrors if calls were made when they are not possible
* ValueErrors if arguments are invalid

last two are callers errors and avoidable.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Use builtin errors when possible

There's on value in using custom errors when builtins work.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Make BadHashError derive from RepositoryError

This is backwards-compatible and means that most (all?) errors resulting
from suspicious or broken metadata are now RepositoryErrors.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
MetadataBundle: Don't do any file IO

Remove file IO from MetadataBundle:
* This make the bundle API very clear and easy to understand
* This means caller must now read from and persist data to disk
  but initial prototypes suggest this won't make Updater too
  complex

This change is something we can still back out from if it turns out to
be the wrong decision: the file-persisting MetadataBundle has been tested
and works fine.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
> But on the other hand, if the caller (`Updater`) handles both access to the file system and the network download, MetadataBundle will implement only the actual metadata verification steps from the client workflow, will be I/O agnostic and will provide a single public method per metadata type (basically the now private `_load_*` methods).

I've been hinking about this and I think I agree with you: this makes the bundle really easy to understand (and makes bundle really do one thing well), and I think the Updater will still be manageable -- it's main function during refresh() is to do I/O, either network or filesystem. I'll do this change as part of this PR

jku commented:
First of all, I've rebased so I can use some fixes from newer experimental-client: those who reviewed already can take commit 
800b088 (MetadataBundle: Fix loads of linting issues) as the previous review point, just rebased. So new changes are 
800b088...f2cff95

The major changes are:
* As the last commit, dropped all file I/O as Teodora suggested (this is the last commit on purpose so it's easy to revert if needed, but I do think it makes sense): this means the API is now really simple: one method per metadata type
* API now consistently raises, no bool return values
* Initial tests were added
* reference time is now recorded at initialization

All of the smaller comments should be handled now. This PR also includes 2 commits of PR #1390 to make things work.

sechkova commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,403 @@
+    update the metadata with the caller making decisions on what is updated.
+    """
+
+    def __init__(self, data: bytes):
```
Can we do something to concretize that data has to be the initial trusted root data?
Making it clear that  `MetadataBundle` is initialized always from a trusted root.
Yet I'm aware that this happens just a couple of lines below so my comment should be taken as a suggestion only.

jku replied:
Yeah, maybe I should do one more commit with better docstring coverage.

jku replied:
:heavy_check_mark:

jku commented:
See https://github.com/jku/tuf/blob/experimental-client-use-bundle/tuf/client_rework/updater_rework.py#L241 for a working example of Updater using this code (I'll make that into a PR once this one is merged).

In this version we end up with following lines-of-code counts for the complete client (as counted by sloccount so comments not counted):
```
client_rework   700
api             500
exceptions.py   120
```
This looks pretty good to me

sechkova reviewed (approved):
I like this simpler version. I left one suggestion so that I don't leave this review blank :)

Commit:
MetadataBundle: Improve docstrings

Document arguments and exceptions, improve prose in general. Remove
mention of local file deletion now that file IO is not done here.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
oops the type hint fix I included in the last commit depended on a python3.9 feature: I removed that, the commit is now just docstrings based onTeodoras last comments

joshuagl commented on tests/test_api.py:
```diff
@@ -496,6 +496,21 @@ def test_delegation_class(self):
+        # empty keys and roles
+        delegations_dict = {"keys":{}, "roles":[]}
+        delegations = Delegations.from_dict(delegations_dict.copy())
+        self.assertEqual(delegations_dict, delegations.to_dict())
```
Note: `assertDictEqual()` gives more meaningful messages when this fails.

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,464 @@
+class MetadataBundle(abc.Mapping):
+    """Internal class to keep track of valid metadata in Updater
+
+    MetadataBundle ensures that the collection of metadata in the bundle valid.
```
```suggestion
    MetadataBundle ensures that the collection of metadata in the bundle is valid.
```

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,464 @@
+
+        Args:
+            data: Trusted root metadata as bytes. Note that this metadata will
+                will only be verified by itself: it is the source of trust for
```
```suggestion
               only be verified by itself: it is the source of trust for
```

MVrachev commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -142,23 +125,16 @@ class MetadataBundle(abc.Mapping):
-    def __init__(self, repository_path: str):
-        """Initialize by loading root metadata from disk"""
-        self._path = repository_path
+    def __init__(self, data: bytes):
```
I would suggest renaming `data` to `root_data` to make it even more descriptive.

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,464 @@
+        self._root_update_finished = False
+
+        # Load and validate the local root metadata
+        # Valid root metadata is required
```
```suggestion
        # Valid initial trusted root metadata is required
```

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,464 @@
+                    "New root is not signed by root", new_root.signed
+                )
+
+            if new_root.signed.version != self.root.signed.version + 1:
```
Per the spec, this comes _after_ verifying with a threshold of keys from the new root metadata. The code is certainly nicer if it happens in this same `if`... do you think we should move it?

jku replied:
You're correct... 

I think I'm happy with it being here (this feels like the spec doing things in the non-optimal order and the code happens to look nicer this way as well) but no strong opinions.

MVrachev reviewed (approved):
Overall, it looks like a pretty good implementation.
I left only one suggestion.
Of course, we will iterate over time on this code, but now it looks good to me!

joshuagl commented on tuf/client_rework/metadata_bundle.py:
```diff
@@ -0,0 +1,464 @@
+        if self.root.signed.is_expired(self.reference_time):
+            raise exceptions.ExpiredMetadataError("New root.json is expired")
+
+        # We skip specification step 5.3.11: deleting timestamp and snapshot
+        # with rotated keys is not needed as they will be invalid, are not
+        # loaded and cannot be loaded
```
This comment feels redundant now that the `MetadataBundle` isn't doing any file handling?

joshuagl reviewed (approved):
The final result here looks good. I noticed a couple of docstring issues and a comment that may not make much sense given the removal of disk and network I/O. I did not do a comprehensive review against the spec, but did notice one place where we do things in a slightly different order than specified. Curious to hear your thoughts about that.

eb648d19bc340ebbe3a60d57eae5ddcff76f6b26 makes me a bit anxious – why do we have `Metadata.to_file()` if we can not rely on it to persist metadata which can be verified? Perhaps a topic for a new Issue?

joshuagl commented:
> [eb648d1](https://github.com/theupdateframework/tuf/commit/eb648d19bc340ebbe3a60d57eae5ddcff76f6b26) makes me a bit anxious – why do we have `Metadata.to_file()` if we can not rely on it to persist metadata which can be verified? Perhaps a topic for a new Issue?

I see we already have https://github.com/theupdateframework/tuf/issues/1392 to capture this. Thanks!

jku commented:
> [eb648d1](https://github.com/theupdateframework/tuf/commit/eb648d19bc340ebbe3a60d57eae5ddcff76f6b26) makes me a bit anxious – why do we have `Metadata.to_file()` if we can not rely on it to persist metadata which can be verified?

This does feel like an oversight in the spec as well: it talks about using canonical forms (and canonical json in particular) for signatures... but hashes aren't talked about (and requiring the implementer to parse canonical json before checking hashes would be wrong  anyway). Even the test files in our repo are not in canonical form WRT whitespace.

On the other hand, it's probably not the end of the world: for the client this is not an issue anyway: writing the original bytes seems a lot safer anyway.

joshuagl commented:
> > [eb648d1](https://github.com/theupdateframework/tuf/commit/eb648d19bc340ebbe3a60d57eae5ddcff76f6b26) makes me a bit anxious – why do we have `Metadata.to_file()` if we can not rely on it to persist metadata which can be verified?
> 
> This does feel like an oversight in the spec as well: it talks about using canonical forms (and canonical json in particular) for signatures... but hashes aren't talked about (and requiring the implementer to parse canonical json before checking hashes would be wrong anyway). Even the test files in our repo are not in canonical form WRT whitespace.

Some clarifications in the spec, or the oft suggested [secondary literature](https://github.com/theupdateframework/specification/issues/91), would probably help here.

> On the other hand, it's probably not the end of the world: for the client this is not an issue anyway: writing the original bytes seems a lot safer anyway.

Agreed. That's what the current client does and what I believe the spec is suggesting when it states what name to download the file as.
